syntax = "proto3";

//import "nanopb.proto";
//option (nanopb_fileopt).packed_struct = true;
//python generator/nanopb_generator.py ../buzzcam/protobuf/message.proto -I ../buzzcam/protobuf
//protoc message.proto --python_out=./ -I=./
/**************/

message PacketHeader {
    uint32 system_uid = 1;
    uint32 ms_from_start = 2;
    uint64 epoch = 3;
}

message SimpleSensorReading{
    uint32 index = 1;
    uint32 timestamp_unix = 2;
    float temperature = 3;
    float humidity = 4;
    float co2 = 5;
    float light_level = 6;
}

/** defined by Bosch */
enum SignalIdentifier {
    UNDEFINED           = 0;
    IAQ                 = 1;
    STATIC_IAQ          = 2;
    CO2_EQ              = 3;
    BREATH_VOC_EQ       = 4;
    RAW_TEMPERATURE     = 6;
    RAW_PRESSURE        = 7;
    RAW_HUMIDITY        = 8;
    RAW_GAS             = 9;
    STABILIZATION_STATUS = 12;
    RUN_IN_STATUS       = 13;
    SENSOR_HEAT_COMPEN_TEMP = 14;
    HEAT_COMPEN_HUMID   = 15;
    GAS_PERCENTAGE      = 21;
} ;

/** defined by Bosch */
enum SensorAccuracy {
    UNRELIABLE          = 0;
    LOW_ACCURACY        = 1;
    MEDIUM_ACCURACY     = 2;
    HIGH_ACCURACY       = 3;
} ;

message SensorReading{
    uint32 packet_index = 1;
    uint32 sample_period = 2;
    
    message Payload {
        fixed64 timestamp_sensor = 1;
        uint64 timestamp_unix = 2;
        uint32 timestamp_ms_from_start = 3;
        float signal = 4;
        uint32 signal_dimensions = 5;
        SignalIdentifier sensor_id = 6;
        SensorAccuracy accuracy = 7;
    }
    repeated Payload payload = 3;
}

message SensorConfig{
    uint32 sample_period_ms = 1;
    bool enable_temperature = 2;
    bool enable_humidity = 3;
    bool enable_gas = 4;
}

message SDCardState{
    bool detected = 1;
    uint64 space_remaining = 2;
    uint64 estimated_remaining_recording_time = 3;
}

message MarkState{
    bool beep_enabled = 1;
    uint32 mark_number = 2;
    uint64 timestamp_unix = 3;
}

message MarkPacket{
    optional string annotation = 1; // throttle max character count (e.g., 50)
}

message DiscoveredDevices{
    uint32 number_of_devices = 1;
    message Device{
        uint32 UID = 1;
        float range = 2;
    }
    repeated Device device = 2;
}

message BatteryState{
    bool charging = 1;
    float voltage = 2;
    optional float percentage = 3;
}

message SystemInfoPacket{
    uint32 number_discovered_devices = 1;
    SimpleSensorReading simple_sensor_reading = 2;
    bool device_recording = 3;
    SDCardState sdcard_state = 4;
    MarkState mark_state = 5;
    BatteryState battery_state = 6;
    optional DiscoveredDevices discovered_devices = 7;
}

enum MicSampleFreq {
    SAMPLE_RATE_16000 = 0;
    SAMPLE_RATE_20500 = 1;
    SAMPLE_RATE_44100 = 2;
    SAMPLE_RATE_48000 = 3;
    SAMPLE_RATE_96000 = 4;
}

enum MicBitResolution{
    BIT_RES_8 = 0;
    BIT_RES_16 = 1;
}

enum CompressionType{
    OPUS = 0;
}

message AudioCompression{
    bool enabled = 1;
    CompressionType compression_type = 2;
    uint32 compression_factor = 3;
}

message AudioConfig{
    bool channel_1 = 1;
    bool channel_2 = 2;
    MicSampleFreq sample_freq = 3;
    MicBitResolution bit_resolution = 4;
    AudioCompression audio_compression = 5;
    float estimated_record_time = 6;
}

message ScheduleConfig{
    bool Sunday = 1;
    bool Monday = 2;
    bool Tuesday = 3;
    bool Wednesday = 4;
    bool Thursday = 5;
    bool Friday = 6;
    bool Saturday = 7;
    uint32 start_hour = 8; 
    uint32 start_minute = 9;
    uint32 stop_hour = 10;
    uint32 stop_minute = 11;
}

message LowPowerConfig{
    bool lowPowerMode = 1;
}

message CameraControl{
    bool pair_with_nearby_cameras = 1;
    bool wakeup_cameras = 2;
    bool capture = 3;
}

message NetworkState{
    uint32 number_of_discovered_devices = 1;
    repeated uint32 discovered_device_UID = 2;
    bool force_rediscovery = 3;
}

message ConfigPacket{
    AudioConfig audio_config = 1;
    repeated ScheduleConfig schedule_config = 2;
    SensorConfig sensor_config = 3;
    LowPowerConfig low_power_config = 4;
    CameraControl camera_control = 5;
    NetworkState network_state = 6;
}

message SpecialFunction{
    oneof payload{
        bool format_sdcard = 1;
        bool function_2 = 2;
        bool function_3 = 3;
        bool function_4 = 4;
        bool function_5 = 5;
        bool function_6 = 6;
    }
}

message Packet {
    PacketHeader header = 1;

    oneof payload {
        SystemInfoPacket system_info_packet = 2;
        MarkPacket mark_packet = 3;
        ConfigPacket config_packet = 4;
        SpecialFunction special_function = 5;
    }
}

